src/main/java/com/aphyr/riemann/client/LocalhostResolver.java

```diff
@@ -0,0 +1,93 @@
+package com.aphyr.riemann.client;
+
+import java.net.UnknownHostException;
+
+/**
+ * A "Smarter" localhost resolver
+ * see issue: https://github.com/aphyr/riemann-java-client/issues/44
+ * Trying to avoid a lot of calls to java.net.InetAddress.getLocalHost()
+ * which under AWS trigger DNS resolving and have relatively high latency *per event*
+ * usually, the hostname doesn't change so often to warrant a real query.
+ *
+ * A real call to java.net.InetAddress.getLocalHost().getHostName()
+ * is made only if:
+ * 1) the refresh interval has passed (=result is stale)
+ * AND
+ * 2) no env vars that identify the hostname are found
+ */
+public class LocalhostResolver {
+
+    // default hostname env var names on Win/Nix
+    public static final String COMPUTERNAME = "COMPUTERNAME"; // Windows
+    public static final String HOSTNAME = "HOSTNAME"; // Nix
+
+    // how often should we refresh the cached hostname
+    public static long refreshIntervalMillis = 60 * 1000;
+
+    // cached hostname result
+    private static boolean envResolved = false;
+    private static String hostname;
+
+    // update (refresh) time management
+    private static long lastUpdate = 0;
+    public static long getLastUpdateTime() { return lastUpdate; }
+    // this is mostly for testing, plus ability to force a refresh
+    public static void setLastUpdateTime(long time) {
+        lastUpdate = time;
+    }
+
+    static {
+        resolveByEnv();
+    }
+
+    /**
+     * get resolved hostname.
+     * encapsulates all lookup and caching logic.
+     *
+     * @return the hostname
+     */
+    public static String getResolvedHostname() {
+        long now = System.currentTimeMillis();
+        if((now - refreshIntervalMillis) > lastUpdate) {
```

Comment:
My understanding of http://cs.oswego.edu/pipermail/concurrency-interest/2012-January/008795.html was that nanoTime goes through a monotonic CAS on Oracle JVMs, but http://bugs.java.com/bugdatabase/view_bug.do?bug_id=6458294 suggests otherwise. However, the bug *also* notes:

"Certainly nanoTime should be no worse than currentTimeMillis() and so should always be preferred for doing timing measurements."